src/GeometricTools_gridspecials.jl

```diff
@@ -547,7 +547,108 @@ function isedge(grid::GridTriangleSurface, ci::Int; whichedge::Int=0)
+                    algorithm=2, areas=nothing)
+
+Converts specified cell-centered field data to node-based data
+by averaging field values of cells surrounding the node.
+Algorithms: 1.Averaging 2.Area-weighted
+"""
+function get_nodal_data(grid::GridTriangleSurface, field_name::String;
+                            algorithm=2, areas=nothing)
 
+    # Create a vector where each index represents a node
+    nodal_data = zeros(grid.nnodes)
+    net_weight = zeros(grid.nnodes)
+
+    # Weight can be 1 or area. 1 implies simple averaging.
+    if algorithm == 2
+        if isnothing(areas)
+            weight = [get_area(grid, i) for i in 1:grid.ncells]
+        end
+    else
+        weight = ones(grid.ncells)
+    end
+
+    # Parse through each cell and add its field value to that nodal array
+    # whose index is the node index. At the end, obtain the weighted average by
+    # dividing each element of the nodal array using the net weight
+    vtxs = ones(Int, 3)
+    for i = 1:grid.ncells
+        vtxs .= get_cell(grid, i)
+        nodal_data[vtxs] .+= weight[i] * grid.field[field_name]["field_data"][i]
+        net_weight[vtxs] .+= weight[i]
+    end
+
+    # Divide each element by number of cells or net area to obtain average
+    return nodal_data ./ net_weight
+end
+
+"""
+    get_nodal_data(grid::GridTriangleSurface, field_vals; algorithm=2, areas=nothing)
+
+Converts specified cell-centered field data to node-based data
+by averaging field values of cells surrounding the node. 
+`field_vals` could be an array or vector containing the scalar data values.
+Algorithms: 1.Averaging 2.Area-weighted
+"""
+function get_nodal_data(grid::GridTriangleSurface, field_vals; algorithm=2, areas=nothing)
+
+    if length(field_vals) != grid.ncells
+        error("No. of field_vals do not match no. of cells")
+    end
+
+    # Create an array where each index represents a node
+    nodal_data = zeros(grid.nnodes)
+    net_weight = zeros(grid.nnodes)
+
+    # Weight can be 1 or area. 1 implies simple averaging.
+    weight = ones(grid.ncells)
+    if algorithm == 2
+        if isnothing(areas)
+            weight = [get_area(grid, i) for i in 1:grid.ncells]
+        end
+    end
+
+    # Parse through each cell and add its field value to that nodal array
+    # whose index is the node index. At the end, obtain the weighted average by
+    # dividing each element of the nodal array using the net weight
+    vtxs = ones(Int, 3)
+    for i = 1:grid.ncells
+        vtxs .= get_cell(grid, i)
+        nodal_data[vtxs] .+= weight[i] * field_vals[i]
+        net_weight[vtxs] .+= weight[i]
+    end
+
+    # Divide each element by number of cells to obtain average
+    return nodal_data ./ net_weight
+end
+
+"""
+    project_3d_2d(p1, p2, p3)
+
+Project 3D vertices of triangle element on to a 2D coordinate system.
+Returns coordinates and basis vectors. First coordinate is always origin [0, 0] and hence not returned.
+"""
+function project_3d_2d(p1, p2, p3)
```

Comment:
I have the feeling that this function will be called a lot, so it might be good to unroll broadcasting to minimize memory allocation:

```julia
function project_3d_2d(p1, p2, p3)
       a1 = p2[1] - p1[1]
       a2 = p2[2] - p1[2]
       a3 = p2[3] - p1[3]
       maga = sqrt(a1^2 + a2^2 + a3^2)

       e11 = a1/maga
       e12 = a2/maga
       e13 = a3/maga
       .
       .
       .
       b_dot_e1 = b1*e11 + b2*e12 + b3*e13
       .
       .
       .
       t2 = [dot_a_e1, 0.0]
       t3 = [b_dot_e1, dot_b_e2]

       return t2, t3, e1, e2
end
````

Reply:
Unrolled and made project_3d_2d!() an in-place fuction to reduce allocations.

Resolved in [f22ecbb](https://github.com/byuflowlab/GeometricTools.jl/pull/5/commits/f22ecbbcf81d49b26ac09fdafcd7bf254459cd97)

Reply:
![Screenshot from 2023-10-06 12-02-58](https://github.com/byuflowlab/GeometricTools.jl/assets/34376424/6765dc81-9d6c-40bc-a986-9b622db8b225)